### PR TITLE
fix(loki sink): Add accept option to out_of_order_action

### PR DESCRIPTION
**Pull request by fpytloun**
Fixes: https://github.com/vectordotdev/vector/issues/9223

### CODE REVIEW

**Comment by spencergilbert** [comment]
I believe in previous discussions we determined that `ignore` was ambivalent as to whether "vector would ignore the event and not send it" or "vector is going to ignore that the timestamp is out of order". What do you think about about `accept` as an alternative.

Also noted in the original issue:

> log lines are accepted only when their timestamp is not earlier than `time_of_most_recent_line - (max_chunk_age/2)`.

so logs can still be rejected from being too old. We believed we'd need to reasonably handle that internally to vector, as well as document it for users (logging the occurrence if it can be detected from loki's response).

**Comment by fpytloun** [comment]
@spencergilbert So should I change the code to match different naming ("accept")? Or is someone going to finish this fix?
From code perspective, I believe my change should be working fine? But I am new to Rust.

**Comment by fpytloun** [comment]
Found hard-coded concurrency limit:
```
        let service = tower::ServiceBuilder::new()
            .concurrency_limit(1)
            .service(self.service);
```
This should probably also go away as now flushes can be concurrent and adaptive concurrency should be also used.

**Comment by jszwedko** [comment]
> Confirmed that Loki returns 400 Bad Request if entry is "too far behind":
> 
> ```
> 400 Bad Request entry with timestamp 2022-02-22 08:50:09.521941661 +0000 UTC ignored, reason: 'entry too far behind' for stream:
> ```

Thanks for confirming 🙂 I think, for the initial implementation here, we can just verify that Vector drops those events and continues.

**Comment by cirias** [comment]
Hi all! I'm also waiting on this feature, and would like to help if needed. Sometime I want to clarify is, to identify which events is "too far behind", `max_chunk_age` is needed. Should we add a new parameter for that?

**Comment by spencergilbert** [comment]
> Hi all! I'm also waiting on this feature, and would like to help if needed. Something I want to clarify is, to identify which events is "too far behind", `max_chunk_age` is needed. Should we add a new parameter for that?

For the initial implementation @cirias it looks like we'll just be dropping those requests rather than trying to keep Vector's understanding of "max age" in sync with Loki.

@fpytloun - I was wondering if you'd had a chance to take a look at the integration tests here, and if you needed to pass this on for someone else to wrap it up?

**Comment by fpytloun** [comment]
> @fpytloun - I was wondering if you'd had a chance to take a look at the integration tests here, and if you needed to pass this on for someone else to wrap it up?

Yes, just pushed test for accept behavior and updated Loki version. Is this fine like this?

**Comment by jszwedko** [comment]
Thanks @fpytloun ! I pushed a couple of commits to make a few adjustments, but overall looks good!

I'll tag another team member for review.

File: src/sinks/loki/integration_tests.rs [hunk 1]
```diff
@@ -275,12 +275,10 @@ async fn many_tenants() {
         .map(Event::from)
         .collect::<Vec<_>>();
 
-    for i in 0..10 {
-        let event = events.get_mut(i).unwrap();
+    for (i, event) in events.iter_mut().enumerate() {
```
Comment (jszwedko) [inline src/sinks/loki/integration_tests.rs:278]: Just spotted this while looking at this file. It is functionally the same.

File: scripts/integration/docker-compose.loki.yml [hunk 2]
```diff
@@ -1,10 +1,8 @@
 services:
 
   loki:
-    image: grafana/loki:${LOKI_VERSION:-2.1.0}
-    command: -config.file=/etc/loki/loki-config.yaml
-    volumes:
-      - ${PWD}/tests/data:/etc/loki
+    image: grafana/loki:${LOKI_VERSION:-2.4.1}
+    command: -config.file=/etc/loki/local-config.yaml -auth.enabled=true
```
Comment (jszwedko) [inline scripts/integration/docker-compose.loki.yml:5]: I moved the `auth.enabled` to a flag here so we can just use the default `local-config.yaml` and delete our custom one which the newer version of Loki was not starting with.

**Comment by jszwedko** [comment]
Thanks @fpytloun I pushed a couple of commits fixing it up a bit. I backed out the concurrency adjustments because I think that needs more work to get right (it should be respecting the `request` options, but it looks like this was accidentally dropped in https://github.com/vectordotdev/vector/pull/9506/). I realize that this option won't really offer too much advantage until we re-enable concurrency, but I'd like to do that as a follow-up PR to keep this one focused.

**Review by spencergilbert — APPROVED** [review]

**Comment by jszwedko** [comment]
I opened https://github.com/vectordotdev/vector/pull/11761 to re-enable concurrency when the `out_of_order_action` is `Accept`.

**Comment by fpytloun** [comment]
Awesome, thank you!

**Pull request merged**